architecture/2025-06-non-k8s-controlplane/2025-06-graph-db-replace-planes.md

```diff
@@ -0,0 +1,489 @@
+2.  **Dual Storage Architecture:** Resources will continue to be stored using the current `database.Client` interface (etcd via Kubernetes APIServer, or Postgres), while application graph relationships will be managed by the GAL with graph database backends.
+3.  **Pluggable Backend Support:** The GAL will support both Kùzu (for dev/test/POC) and Postgres with Apache AGE (for production), with configuration to select the backend.
+4.  **Schema Definition:** A formal schema for application graph entities (Applications, Environments, Resources as nodes) and their relationships (as edges with types and properties) will be defined and enforced in both graph backends.
+5.  **Minimal Graph Data Storage:** The GAL will store only essential resource metadata in the graph database (resource ID, type, connections, and optional properties needed for query filtering), with full resource data retrieval continuing through the existing `database.Client` interface based on graph query results.
+6.  **Data Synchronization:** The GAL will maintain synchronization between resource changes (via `database.Client`) and their corresponding minimal graph representations, ensuring the application graph accurately reflects the current state of resource relationships and queryable properties.
+7.  **Component Updates:** Radius components that currently perform application graph operations (traversals, relationship queries) will be updated to use the GAL for graph queries, then retrieve full resource data via `database.Client` based on the graph results, while continuing to use `database.Client` directly for individual resource CRUD operations.
+
+**Storage Strategy Rationale:** We chose to maintain the existing `database.Client` interface for resource storage because: (1) The current Postgres and etcd backends already provide excellent performance for individual resource operations; (2) Graph databases are optimized for relationship queries, not necessarily single-record lookups; (3) This approach allows incremental migration with lower risk; (4) Different data types (individual resources vs. relationships) have different access patterns and requirements; (5) Production deployments can continue using battle-tested storage patterns while gaining graph capabilities for application relationship queries. **The GAL will store only minimal resource metadata (ID, type, connections, and optional query-filtering properties) in the graph database, with complete resource data retrieval continuing through the existing `database.Client` based on graph query results.**
+
+#### Architecture Diagram
+
+```mermaid
+graph TD
+    A["Radius CLI/API <br/>(User Interactions)"]
+    B["Radius CoreRP<br/>"]
+    C["database.Client <br/>(Resource Storage)"]
+    D["Graph Access Layer <br/>(Application Graph)"]
+    E1["etcd (Current) | Postgres <br/>(Production)"]
+    F1["Kùzu Embedded DB <br/>(Dev/Test/POC)"]
+    F2["Postgres + Apache AGE <br/>(Production)"]
```

Comment:
I recommend collapsing these two nodes into a single node that is something like "Cypher compliant graph DB", unless we plan to hard code the dependency on Kuzu and Apache.